### PR TITLE
GH-40799: [Doc][Format] Implementation status page should list canonical extension types

### LINKED ISSUE #40799: [Doc][Format] Implementation status page should list canonical extension types
### Describe the enhancement requested

Canonical extension types should be listed as a new subsection at the end of the https://arrow.apache.org/docs/dev/status.html#data-types section.

### Component(s)

Documentation, Format

### PR DESCRIPTION
**Pull request by rok**
### Rationale for this change

Two specifications and one implementation of canonical extension types were added and this should be documented.

### What changes are included in this PR?

This represents current state of canonical extension types.

### Are these changes tested?

No, docs only.

### Are there any user-facing changes?

In so much as they read docs.
* GitHub Issue: #40799

### CODE REVIEW

**Comment by rok** [comment]
@github-actions crossbow submit preview-docs

**Comment by rok** [comment]
[Current docs](https://arrow.apache.org/docs/dev/status.html#data-types), [proposed docs](http://crossbow.voltrondata.com/pr_docs/41053/status.html#data-types)

**Review by jorisvandenbossche — APPROVED** [review]

File: docs/source/status.rst [hunk 1]
```diff
@@ -111,14 +111,24 @@ Data Types
 | Run-End Encoded   | ✓     |       | ✓     |            |       |       |       |       |
 +-------------------+-------+-------+-------+------------+-------+-------+-------+-------+
 
++---------------------+-------+-------+-------+------------+-------+-------+-------+-------+
+| Canonical           | C++   | Java  | Go    | JavaScript | C#    | Rust  | Julia | Swift |
+| Extension types     |       |       |       |            |       |       |       |       |
++=====================+=======+=======+=======+============+=======+=======+=======+=======+
+| FixedShapeTensor    | ✓     |       |       |            |       |       |       |       |
++---------------------+-------+-------+-------+------------+-------+-------+-------+-------+
+| VariableShapeTensor |       |       |       |            |       |       |       |       |
```
Comment (pitrou) [inline docs/source/status.rst:120]: These are the C++ and Python internal names, should we use the prose spelling instead? "Fixed shape tensor" and "Variable shape tensor"

**Review by pitrou — COMMENTED** [review]
Thanks a lot @rok , just a nit.

File: docs/source/status.rst (hunk 1 above)
Comment (rok) [inline docs/source/status.rst:120]: Good point, thanks for catching this!
```suggestion
| Fixed shape tensor    | ✓     |       |       |            |       |       |       |       |
+---------------------+-------+-------+-------+------------+-------+-------+-------+-------+
| Variable shape tensor |       |       |       |            |       |       |       |       |
```

File: docs/source/status.rst [hunk 2]
```diff
@@ -111,14 +111,24 @@ Data Types
...
 +-------------------+-------+-------+-------+------------+-------+-------+-------+-------+
 
++---------------------+-------+-------+-------+------------+-------+-------+-------+-------+
+| Canonical           | C++   | Java  | Go    | JavaScript | C#    | Rust  | Julia | Swift |
+| Extension types     |       |       |       |            |       |       |       |       |
++=====================+=======+=======+=======+============+=======+=======+=======+=======+
+| Fixed shape tensor    | ✓     |       |       |            |       |       |       |       |
++---------------------+-------+-------+-------+------------+-------+-------+-------+-------+
+| Variable shape tensor |       |       |       |            |       |       |       |       |
++---------------------+-------+-------+-------+------------+-------+-------+-------+-------+
```
Comment (rok) [inline docs/source/status.rst:121]: Fixing line lengths.
```suggestion
+-----------------------+-------+-------+-------+------------+-------+-------+-------+-------+
| Canonical             | C++   | Java  | Go    | JavaScript | C#    | Rust  | Julia | Swift |
| Extension types       |       |       |       |            |       |       |       |       |
+=======================+=======+=======+=======+============+=======+=======+=======+=======+
| Fixed shape tensor    | ✓     |       |       |            |       |       |       |       |
+-----------------------+-------+-------+-------+------------+-------+-------+-------+-------+
| Variable shape tensor |       |       |       |            |       |       |       |       |
+-----------------------+-------+-------+-------+------------+-------+-------+-------+-------+
```

File: docs/source/status.rst (hunk 1 above)
Comment (rok) [reply docs/source/status.rst]: Good point, thanks for catching this! Fixed.

**Review by pitrou — APPROVED** [review]

**Pull request merged**